applications/FluidDynamicsApplication/python_scripts/navier_stokes_two_fluids_hydraulic_solver.py

```diff
@@ -0,0 +1,692 @@
+
+        #FIXME: Check that this is required. Note that right now the convection process is the one computing NODAL_H if nodal tau is active
+        find_nodal_h = KratosMultiphysics.FindNodalHNonHistoricalProcess(self.GetComputingModelPart())
+        find_nodal_h.Execute()
+
+        KratosMultiphysics.Logger.PrintInfo(self.__class__.__name__, "Solver initialization finished.")
+
+
+    def InitializeSolutionStep(self):
+
+        # Inlet and outlet water discharge is calculated for current time step, first discharge and the considering the time step inlet and outlet volume is calculated
+        if self.mass_source:
+            self._ComputeInitialStepWaterVolume()
+
+        # Perform the level-set convection according to the previous step velocity
+        self._PerformLevelSetConvection()
+        KratosMultiphysics.Logger.PrintInfo(self.__class__.__name__, "Level-set convection is performed.")
+
+        # Perform distance correction to prevent ill-conditioned cuts
+        self._GetDistanceModificationProcess().ExecuteInitializeSolutionStep()
+
+        # Update the DENSITY and DYNAMIC_VISCOSITY values according to the new level-set
+        self._SetNodalProperties()
+
+        # Accumulative water volume error ratio due to level set. Adding source term
+        self._ComputeVolumeError()
+
+        # Perform the convection of the historical database (Eulerian FM-ALE)
+        if self.eulerian_fm_ale:
+            self._PerformEulerianFmAleVelocity()
+
+        # Initialize the solver current step
+        self._GetSolutionStrategy().InitializeSolutionStep()
+
+        # We set this value at every time step as other processes/solvers also use them
+        # Note that this is required as the convection processes may set a different value (this is the one to be used in the Navier-Stokes element)
+        dynamic_tau = self.settings["formulation"]["dynamic_tau"].GetDouble()
+        self.main_model_part.ProcessInfo.SetValue(KratosMultiphysics.DYNAMIC_TAU, dynamic_tau)
+
+    def FinalizeSolutionStep(self):
+        KratosMultiphysics.Logger.PrintInfo(self.__class__.__name__, "Mass and momentum conservation equations are solved.")
+
+        # Recompute the distance field according to the new level-set position
+        if self._reinitialization_type != "none":
+            self._GetDistanceReinitializationProcess().Execute()
+            KratosMultiphysics.Logger.PrintInfo(self.__class__.__name__, "Redistancing process is finished.")
+
+        # Prepare distance correction for next step
+        self._GetDistanceModificationProcess().ExecuteFinalizeSolutionStep()
+
+        # Finalize the solver current step
+        self._GetSolutionStrategy().FinalizeSolutionStep()
+
+        # Acceleration for generalised alpha time integration method.
+        self.__CalculateTimeIntegrationAcceleration()
```

Comment:
I might be wrong, but shouldn't we do this in the `FinalizeSolutionStep`? I think that in the no-predict case as we do now it makes no difference but wouldn't be the case if we add a prediction.

Reply:
After a discussion we decided to do it in the FinalizeSolutiolStep